web3/providers/persistent.py

```diff
@@ -197,6 +200,9 @@ async def _match_response_id_to_request_id() -> RPCResponse:
                     )
                     return popped_response
                 else:
+                    # check if an exception was recorded in the listener task and raise
```

Comment:
This change allows for getting the response which has already been received before handling the message listener exceptions.

As shown in the test change, the server sends a series of responses and closes the connection (which raises an `ConnectionClosedOK`); before this change, although the client had received the responses in the buffer (`_request_response_cache`), it did not return the response, but instead raised a "ConnectionClosed" error.

As a reference, we can see that the ethers.js [implementation](https://github.com/ethers-io/ethers.js/blob/fc66b8ad405df9e703d42a4b23bc452ec3be118f/src.ts/providers/provider-socket.ts#L309) prioritizes the return of a response before handling the exception.

Reply:
Regardless of the test change, which I don't believe needs to happen if we refactor the message queue in the mock class a bit, I think this change is ideal so we should keep it.